compiler/resolution/fixupExports.cpp

```diff
@@ -0,0 +1,425 @@
+      USR_FATAL(fn, "Return type %s of exported routine %s may only have "
+                    "return intent %s",
+                t->name(), fn->userString,
+                retTagDescrString(RET_VALUE));
+    }
+  }
+
+  return;
+}
+
+static FnSymbol* createWrapper(FnSymbol* fn) {
+  FnSymbol* result = fn->copy();
+  fn->defPoint->insertAfter(new DefExpr(result));
+  // TODO: Is there a better way to empty the block body?
+  result->body->replace(new BlockStmt());
+  INT_ASSERT(fn->hasFlag(FLAG_EXPORT));
+  fn->removeFlag(FLAG_EXPORT);
+  return result;
+}
+
+static Type* getTypeForFixup(Type* t, bool ret) {
+  if (t == dtString) {
+    Type* result = ret ? getCharPtrType() : dtStringC;
+    return result;
+  } else {
+    INT_FATAL("Unsupported type for formal in: %s", __FUNCTION__);
+  }
+
+  // Should never reach here.
+  return NULL;
+}
+
+//
+// TODO: Can insert non-copying versions when we are doing multi-locale
+// interop.
+//
+static const char* getConversionCallName(Type* srctype, Type* dsttype) {
+  const char* result = NULL;
+
+  // TODO: We should probably maintain this in a table or the like.
+  if (srctype == dtString && dsttype == dtStringC) {
+    result = "chpl__exportStringToConstCharPtr";
+  } else if (srctype == dtStringC && dsttype == dtString) {
+    result = "chpl__exportConstCharPtrToString";
+  } else if (srctype == dtString && dsttype == getCharPtrType()) {
+    result = "chpl__exportStringToCharPtr";
+  } else if (srctype == getCharPtrType() && dsttype == dtString) {
+    result = "chpl__exportCharPtrToString";
+  } else {
+    INT_FATAL("Bad types for conversion call in: %s", __FUNCTION__);
+  }
+
+  return result;
+}
+
+static FnSymbol* getConversionCall(Type* srctype, Type* dsttype) {
+  const char* name = getConversionCallName(srctype, dsttype);
+  INT_ASSERT(name != NULL);
+
+  std::map<const char*, FnSymbol*>& m = conversionCallMap;
+  std::map<const char*, FnSymbol*>::iterator it;
+  it = m.find(name);
+  if (it != m.end()) { return it->second; }
+
+  FnSymbol* result = resolveConversionCall(srctype, dsttype);
+  m[name] = result;
+
+  return result;
+}
+
+//
+// TODO: This assumes that the names of the conversion calls are unique and
+// _are not_ being overloaded.
+//
+static FnSymbol* resolveConversionCall(Type* srctype, Type* dsttype) {
+  const char* name = getConversionCallName(srctype, dsttype);
+
+  BlockStmt* block = new BlockStmt();
+  VarSymbol* tmp = newTemp(srctype);
+  CallExpr* call = new CallExpr(name, tmp);
+  FnSymbol* result = NULL;
+ 
+  // Modifications to `chpl_gen_main` are only temporary.
+  chpl_gen_main->body->insertAtHead(block);
+
+  block->insertAtTail(new DefExpr(tmp));
+  block->insertAtTail(call);
+
+  tryResolveCall(call);
+
+  result = call->resolvedFunction();
+  INT_ASSERT(result != NULL);
+
+  //
+  // These are not called by user code paths, so we must resolve them now or
+  // we will get AST errors during late stage checking.
+  //
+  normalize(result);
```

Comment:
Does this really need to be normalized?

Reply:
I can always try to remove the normalization call!